File: typescript/facade/facade.test.ts
import { CheckingAccount } from "./account/checking-account";
import { Bank } from "./bank";
import { ConsoleLog } from "./log/console-log";
import { Notification } from "./notification/notification";

test('Test Facade', () => {
    const bank = new Bank(
        new CheckingAccount(10_000_000),
        new ConsoleLog(),
        new Notification()
    );

    bank.depositWithLogAndNotification(5000);
    expect(bank.getBalance()).toEqual(10_005_000);

    bank.withdrawWithLogAndNotification(5000);
    expect(bank.getBalance()).toEqual(10_000_000);

    expect(() => bank.depositWithLogAndNotification(5_000_001)).toThrow(Error)

    expect(() => bank.withdrawWithLogAndNotification(bank.getBalance() + 1)).toThrow(Error)
});

File: go/facade/facade_test.go
package facade

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestFacade(t *testing.T) {
	t.Run("Test Facade", func(t *testing.T) {
		bank := NewBank(
			NewCheckingAccount(10_000_000),
			NewConsoleLog(),
			NewNotification(),
		)

		err := bank.DepositWithLogAndNotification(5000)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 10_005_000, bank.GetBalance())

		err = bank.WithdrawWithLogAndNotification(5000)
		if err != nil {
			t.Error(err)
		}
		assert.Equal(t, 10_000_000, bank.GetBalance())

		err = bank.DepositWithLogAndNotification(5_000_001)
		assert.NotNil(t, err)
		assert.NotEmpty(t, err.Error())

		err = bank.WithdrawWithLogAndNotification(bank.GetBalance() + 1)
		assert.NotNil(t, err)
		assert.NotEmpty(t, err.Error())
	})
}
